### PR TITLE
feat: add AtRustAttr to generate_struct

### PR DESCRIPTION
**Pull request by Mehrbod2002**
Fixes: #2660 
Related: #2727 
The Test added too.

Implement new feature `@pragma` and use `serde` as keyword ( also can add new feature with it ) and attach Serialize and Deserialize to `generate_struct` .

Eaxmple: 
```javascript
@pragma(serde)
export struct Test {
   foo: int,
}
```

### CODE REVIEW

File: internal/compiler/parser.rs [hunk 1]
```diff
@@ -322,7 +322,7 @@ declare_syntax! {
     }
     // syntax kind
     {
-        Document -> [ *Component, *ExportsList, *ImportSpecifier, *StructDeclaration ],
+        Document -> [ *Component, *ExportsList, *ImportSpecifier, *StructDeclaration, *AtPragma ],
```
Comment (tronical) [inline internal/compiler/parser.rs:325]: It seems to me that this is the wrong location in the "AST". I think it should be attached (as child) to those elements where it applies, i.e. initially just structure declarations.

**Review by tronical — COMMENTED** [review]
It's tempting to go with `@pragma(serde)` and add basically built-in support for the serde crate. It's a kind of de-facto standard and that kind of syntax is really convenient.

I admit I'm also a little more attracted by Olivier's [rust-attr](https://github.com/slint-ui/slint/pull/2727#issuecomment-1549300108) idea that's more generic. I see two advantages over a `@pragma(serde)`:

1. It results in more readable .slint code: When reading you know that this is rust specific. It doesn't apply to C++, and `@pragma(serde)` will never apply to C++.
2. It works with other crates and derive macros as well.

Yes, it's less convenient, but my preference is the `rust-attr` idea.

That said, I'll provide some feedback on this patch.

**Comment by Mehrbod2002** [comment]
@tronical Agree .
I am confused, a thing about rust-attr , is it Syntax which should be enable like pragma ? Because you mentioned it is less convinced problem .

**Comment by tronical** [comment]
`@rust-attr(cfg_attr(feature="serde", derive(Serialize, Deserialize)))` is more verbose than `@pragma(serde)`, but it is also more generic and - despite the verbosity - more readable, because it conveys two critical aspects to the reader:

1. The feature only affects the Rust generated code.
2. This particular example just injects a cfg_attr - so it's transparent to the reader what exactly this does.

This is in contrast to `@pragma(serde)`, where the information that this is (a) rust specific and (b) how exactly it relates to the serde crate is located somewhere completely different (deep in the Slint documentation).

Does this make sense? :)

**Comment by Mehrbod2002** [comment]
@tronical Thanks a lot for information . I am agree with the idea . if you let me to validate it ( can I ? :) )  ,
so you mean the snipp code will look like this ?? 
```javascript
@rust-attr(cfg_attr(feature="serde", derive(Serialize, Deserialize)))
export struct Test {
   foo: int,
}
``` .
so will slint parse it like `@pragma` ?

File: internal/compiler/parser.rs (hunk 1 above)
Comment (Mehrbod2002) [reply internal/compiler/parser.rs:325]: But as Struct is Document element , I suppose AtPragma according to its syntax it should be an specif Document element which contains Struct . isn't it ?

**Comment by tronical** [comment]
Yes. Instead of supporting `@pragma` in the parser, the parser would learn to parse `@rust-attr(...)`.

So `@rust-attr(XXX)` would become `#[XXX]` in Rust generated code.

I see though that the hard part is making this work in the parser... It appears before `export` in the CST, but it really belongs to the `struct` in the AST.

**Comment by Mehrbod2002** [comment]
> I see though that the hard part is making this work in the parser... It appears before `export` in the CST, but it really belongs to the `struct` in the AST.

This part is already done . `@pragma` works with default struct and export struct . it will start node and detect export keyword in its parser and also as it is document element it will extend to export list too .

**Comment by Mehrbod2002** [comment]
@tronical 
New one pushed .
I don't know any about feature if use cfg feature on macro too , can i ask you to commit this section ? Add feature to toml and then use cfg feature instead default derive in `rust.rs`

File: internal/compiler/Cargo.toml [hunk 2]
```diff
@@ -53,6 +53,7 @@ once_cell = "1"
 url = "2.2.1"
 dunce = "1.0.1"
 linked_hash_set = "0.1.4"
+serde = { version = "1.0.163", features = ["derive"] }
```
Comment (ogoffart) [inline internal/compiler/Cargo.toml:56]: I don't think we need this.  Why is that required?

File: internal/compiler/generator/cpp.rs [hunk 3]
```diff
@@ -579,7 +579,8 @@ pub fn generate(doc: &Document) -> impl std::fmt::Display {
     }
 
     for ty in doc.root_component.used_types.borrow().structs.iter() {
-        if let Type::Struct { fields, name: Some(name), node: Some(node) } = ty {
+        // FIXME: add feature as deriven
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:582]: I think it's fine to just ignore this in C++ since there is no such things there.  So no need for this comment

File: internal/compiler/parser.rs [hunk 4]
```diff
@@ -423,7 +423,10 @@ declare_syntax! {
         ArrayType -> [ Type ],
         /// `struct Foo := { ... }
         StructDeclaration -> [DeclaredIdentifier, ObjectType],
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply internal/compiler/parser.rs:425]: I agree, I think it would be simpler for consumer of the tree if StructDeclaration contained the `AtRustAttr`

File: internal/compiler/parser/document.rs [hunk 5]
```diff
@@ -45,6 +46,35 @@ pub fn parse_document(p: &mut impl Parser) -> bool {
                     break;
                 }
             }
+            "@" => {
+                let mut is_export = false;
```
Comment (ogoffart) [inline internal/compiler/parser/document.rs:50]: So you can add a checkpoint in the parser here and parse the `@rust-attr(...)` node here, and then start the struct node at that checkpoint.
So that the AtRustAttr node would be within the struct

File: internal/compiler/parser/type.rs [hunk 6]
```diff
@@ -85,3 +85,89 @@ pub fn parse_struct_declaration(p: &mut impl Parser) -> bool {
...
+macro_rules! deriven_macro_trait {
+    ($feature:expr) => {
+        match $feature {
+            "serde" => vec!["Serialize", "Deserialize"],
+            _ => vec![],
+        }
+    };
+}
+
+pub fn parse_deriven(p: &mut impl Parser) -> bool {
```
Comment (ogoffart) [inline internal/compiler/parser/type.rs:121]: I would actually just treat the inside of a @rust-attr  as some opaque token, and just have these token be replicated verbatim in the generated rust.

This has the advantage of being generic, but has the problem that compilation errors might be harder to track to their source.  But I think for this kind of feature, we really want to support anything.
Such as `@rust-attr(derive(serde::Serialize, serde::Deserialize))`  or even `@rust-attr(doc = "whatever")`

File: internal/compiler/langtype.rs [hunk 7]
```diff
@@ -60,6 +60,8 @@ pub enum Type {
         name: Option<String>,
         /// When declared in .slint, this is the node of the declaration.
         node: Option<syntax_nodes::ObjectType>,
+        /// deriven
+        feature: Option<Vec<String>>,
```
Comment (ogoffart) [inline internal/compiler/langtype.rs:64]: Maybe rust_attributes would be a better name?

File: internal/compiler/load_builtins.rs [hunk 8]
```diff
@@ -34,9 +34,13 @@ pub fn load_builtins(register: &mut TypeRegister) {
     let doc: syntax_nodes::Document = node.into();
 
     // parse structs
-    for s in doc.StructDeclaration().chain(doc.ExportsList().flat_map(|e| e.StructDeclaration())) {
+    for s in doc
+        .StructDeclaration()
+        .chain(doc.ExportsList().flat_map(|e| e.StructDeclaration()))
+        .chain(doc.ExportsList().flat_map(|e| e.RustAttr().flat_map(|x| x.StructDeclaration())))
```
Comment (ogoffart) [inline internal/compiler/load_builtins.rs:40]: You can safely assume that the builtin file don't have `@rust-attr`

File: internal/compiler/parser/type.rs (hunk 6 above)
Comment (Mehrbod2002) [reply internal/compiler/parser/type.rs:121]: How can we validate it ?

Comment (ogoffart) [reply internal/compiler/parser/type.rs:121]: > How can we validate it ? Should we validate it ? 

I think it can just be an arbitrary balanced token tree, so we can just validate that it has the right amount of parentheses and braces.

> And other thing is it would be used as String , can we use String as evaluate Statement in quote! ?

`TokenStream` implements `FromStr`

**Comment by Mehrbod2002** [comment]
@ogoffart now there are generic rust-attr and move RustAttr to StructDeclaration.

**Comment by Mehrbod2002** [comment]
@ogoffart now there are generic rust-attr and move RustAttr to StructDeclaration.

**Comment by Mehrbod2002** [comment]
@ogoffart can we disable doc tests for StructDeclaration ?

File: internal/compiler/parser.rs [hunk 9]
```diff
@@ -422,8 +422,9 @@ declare_syntax! {
         /// `[ type ]`
         ArrayType -> [ Type ],
         /// `struct Foo := { ... }
-        StructDeclaration -> [DeclaredIdentifier, ObjectType],
-
+        StructDeclaration -> [DeclaredIdentifier, ObjectType, AtRustAttr],
```
Comment (ogoffart) [inline internal/compiler/parser.rs:425]: i believe it should be

```suggestion
        StructDeclaration -> [DeclaredIdentifier, ObjectType, ?AtRustAttr],
```

To specify that this is optional

File: internal/compiler/parser/type.rs [hunk 10]
```diff
@@ -85,3 +85,61 @@ pub fn parse_struct_declaration(p: &mut impl Parser) -> bool {
...
+    parse_type_object(&mut *p);
+    true
+}
+
+fn parse_parentheses(p: &mut impl Parser) -> bool {
+    let mut p = p.start_node(SyntaxKind::AtRustAttr);
+    let mut opened = 0;
+    let mut closed = 0;
+    while closed <= opened {
+        if p.peek().as_str() == "(" {
```
Comment (ogoffart) [inline internal/compiler/parser/type.rs:123]: Instead of using as_str(),   you can use the `kind() == SyntaxKind::LParent`

File: internal/compiler/parser/type.rs [hunk 11]
```diff
@@ -85,3 +85,61 @@ pub fn parse_struct_declaration(p: &mut impl Parser) -> bool {
...
+    let checkpoint = p.checkpoint();
+    debug_assert_eq!(p.peek().as_str(), "@");
+    p.consume(); // "@"
+    if p.peek().as_str() != "rust-attr" {
+        p.expect(SyntaxKind::AtRustAttr);
+    }
+    p.consume(); // "rust-attr"
+    p.expect(SyntaxKind::LParent);
+    parse_parentheses(&mut *p);
+    if p.peek().as_str() == "export" {
```
Comment (ogoffart) [inline internal/compiler/parser/type.rs:99]: We also need to start an ExportList node in that case.

Comment (Mehrbod2002) [reply internal/compiler/parser/type.rs:99]: It starts in document.rs when it reaches to "@" .

File: internal/compiler/parser/document.rs [hunk 12]
```diff
@@ -45,6 +46,35 @@ pub fn parse_document(p: &mut impl Parser) -> bool {
                     break;
                 }
             }
+            "@" => {
```
Comment (ogoffart) [inline internal/compiler/parser/document.rs:49]: I think we should make sure here that we are followed with `rust-attr` 
Otherwise this is a bit fragile and 
`@foobar` will get into this when we don't really what to.  Also the loop is quite fragile.
IMHO, we should do really parse the @rust-attr(...) there and make sure we have the right amount of ballanced parentheses in there and then use the checkpoint to start the StructDeclaration node.

File: internal/compiler/parser/type.rs (hunk 11 above)
Comment (ogoffart) [reply internal/compiler/parser/type.rs:99]: Ah indeed

File: internal/compiler/parser/type.rs [hunk 13]
```diff
@@ -85,3 +85,61 @@ pub fn parse_struct_declaration(p: &mut impl Parser) -> bool {
...
+        if p.peek().kind() == SyntaxKind::LParent {
+            opened += 1;
+        }
+        if p.peek().kind() == SyntaxKind::RParent {
+            closed += 1;
+        }
+        if closed == opened
+            && opened != 0
+            && closed != 0
+            && p.peek().kind() != SyntaxKind::Comma
```
Comment (ogoffart) [inline internal/compiler/parser/type.rs:132]: Why do you check for a coma?

Note, you can also see https://github.com/slint-ui/slint/blob/0c64312383bfea39c4b01d518619bbde2c2c7062/tools/lsp/server_loop.rs#L901  for some example of code that tries to find balanced parentheses

**Review by ogoffart — COMMENTED** [review]
Nice progress.

File: internal/compiler/parser/type.rs (hunk 13 above)
Comment (Mehrbod2002) [reply internal/compiler/parser/type.rs:132]: I suppose it will parse if check comma between element of rust_attr . as we have extra RPRANT which is for rust_attr e can use it . but if we want to check for all possibilty in rust-attr it worth to validate as rust too and you mentioned e can just check it out . it will be big a problem to return error or annouce which element missed . as we want to relay on rust compile we relay on it for this .

**Comment by Mehrbod2002** [comment]
@ogoffart can you consider to review it for failed test /

**Comment by Mehrbod2002** [comment]
it is not workable ? @ogoffart

**Review by ogoffart — COMMENTED** [review]
Change looks almost good, thanks.

The missing thing now is the documentation.
I believe it should be documented in api/rs/slint/lib.rs under `## Type Mappings`

**Review by ogoffart — APPROVED** [review]
